python/services/dataservice/dmod/dataservice/models.py

```diff
@@ -0,0 +1,254 @@
+from pydantic import (
+    BaseModel,
+    root_validator,
+    validator,
+    StrictInt,
+    StrictFloat,
+    StrictStr,
+    Field,
+    UUID4,
+)
+from minio.datatypes import Object
+
+from dmod.core import meta_data
+from .errors import Error as ErrorEnum
+
+from typing import Dict, Optional, Union, List, Generic, TypeVar
+from typing_extensions import Self
+
+T = TypeVar("T")
+
+
+class Option(GenericModel, Generic[T]):
+    # root_validator sets this field
+    is_none: bool = True
+    value: Optional[T]
+
+    @root_validator()
+    def _enforce_none_variant(
+        cls, values: Dict[str, Union[bool, T]]
+    ) -> Dict[str, Union[bool, T]]:
+        if values["value"] is None:
+            values["is_none"] = True
+        else:
+            values["is_none"] = False
+        return values
+
+
+class Error(BaseModel):
+    type: str
+    title: str
+    status: int
+    detail: str
+
+    @classmethod
+    def from_error_enum(
+        cls, error_enum: ErrorEnum, detail: Optional[str] = None
+    ) -> Self:
+        return cls(
+            title=error_enum.name,
+            type=error_enum.type,
+            status=error_enum.status,
+            detail=detail if detail is not None else error_enum.detail,
+        )
+
+
+class StandardDatasetIndex(str, Enum):
+    UNKNOWN = "UNKNOWN"
+    TIME = "TIME"
+    CATCHMENT_ID = "CATCHMENT_ID"
+    DATA_ID = "DATA_ID"
+    HYDROFABRIC_ID = "HYDROFABRIC_ID"
+    LENGTH = "LENGTH"
+    GLOBAL_CHECKSUM = "GLOBAL_CHECKSUM"
+    ELEMENT_ID = "ELEMENT_ID"
+    REALIZATION_CONFIG_DATA_ID = "REALIZATION_CONFIG_DATA_ID"
+
+    def as_core_standard_dataset_index(self) -> meta_data.StandardDatasetIndex:
+        return meta_data.StandardDatasetIndex.get_for_name(self.name)
+
+
+class DataFormat(str, Enum):
+    AORC_CSV = "AORC_CSV"
+    NETCDF_FORCING_CANONICAL = "NETCDF_FORCING_CANONICAL"
+    NETCDF_AORC_DEFAULT = "NETCDF_AORC_DEFAULT"
+    NGEN_OUTPUT = "NGEN_OUTPUT"
+    NGEN_REALIZATION_CONFIG = "NGEN_REALIZATION_CONFIG"
+    NGEN_GEOJSON_HYDROFABRIC = "NGEN_GEOJSON_HYDROFABRIC"
+    NGEN_PARTITION_CONFIG = "NGEN_PARTITION_CONFIG"
+    BMI_CONFIG = "BMI_CONFIG"
+    NWM_OUTPUT = "NWM_OUTPUT"
+    NWM_CONFIG = "NWM_CONFIG"
+    NGEN_CAL_OUTPUT = "NGEN_CAL_OUTPUT"
+    NGEN_CAL_CONFIG = "NGEN_CAL_CONFIG"
+
+    def as_core_data_format(self) -> meta_data.DataFormat:
+        return meta_data.DataFormat.get_for_name(self.name)
+
+
+class DataCategory(str, Enum):
+    CONFIG = "CONFIG"
+    FORCING = "FORCING"
+    HYDROFABRIC = "HYDROFABRIC"
+    OBSERVATION = "OBSERVATION"
+    OUTPUT = "OUTPUT"
+
+    def as_core_data_category(self) -> meta_data.DataCategory:
+        return meta_data.DataCategory.get_for_name(self.name)
+
+
+def _validate_is_not_unknown(value: StandardDatasetIndex):
+    if value == StandardDatasetIndex.UNKNOWN:
+        raise ValueError("unknown restriction variable not allowed")
+    return value
+
+
+class ContinuousRestriction(BaseModel):
```

Comment:
I have similar thoughts about this class, although it is a little different here because it could extend the type in `dmod.core`, which can't be done with the enums.  

I am wondering now if I've forgotten something we discussed, where we'd planned to add these temporarily while we finish the review and integration of the Pydantic changes branch ...

Reply:
Yeah, I dont think we'd explicitly talked about adding in these temporary structures while the pydantic refactor was still underway. Sorry for not leaving notice about that. I should have at least made a note that these structures are temporary.